### DISCUSSION TITLE
How to solve the gradient of the Userobject object and introduce it into the kernel ConservativeAdvection？

### DISCUSSION THREAD

**Discussion by CIACHD** [idaholab/moose#27033] [Q&A General]
I am a newbie in the field of MOOSE and C+. 
After importing the Userobject into the kernel object, how do I find the gradient of the object?
'''params.addRequiredParam<UserObjectName>("convolution", "RadialGreensConvolution user object");
...
 _convolution(getUserObject<RadialGreensConvolution>("convolution").getConvolution())
'''
The governing equation is ∇∙(n * ∇C),  n is Density field to be solved， C is scalar field associated with n，Get its value using RadialGreensConvolution from module magpie。 

I looked at the GenericFunctorGradientMaterial.h/C file and Function.h/C. I don’t know if I can directly use the gradient() method to get the gradient. If so, how should I do it?

**Comment by GiudGiud** [comment]
Hello

A user object isn't a field. You could implement a gradient routine on it but it s probably not the right option. 

Is the field you want the gradient for a variable? A material property ?

Guillaume

**Reply by CIACHD** [reply]
Thanks for your reply
 It is a material property calculated from variables。
Do you mean that I can convert the userobject object into material and then calculate the gradient of material?

##in the Magpie 
![image](https://github.com/idaholab/moose/assets/105799938/2058c18c-b771-421c-91d6-eafc4d395f3b)
the property is calculated by UserObject-RadialGreensConvolution。 I want to calculate the gradient of convolution and and introduce the results into the kernel

**Reply by GiudGiud** [reply]
Hello

I would add a new routine to this user object that computed this gradient

Then in the kernel you use getUserObject (a fe_problem routine) templated with the specific type of this user object to retrieve a pointer to this user object. From this pointer you can call the gradient routine

Guillaume

**Reply by CIACHD** [reply]
Thank you so much！！！

**Reply by CIACHD** [reply]
Hello,Dr.Guillaume
I want to know how long it will take to complete the work.
 I need to use this gradient to further develop my kernel
taozhang

**Reply by GiudGiud** [reply]
If you are comfortable with C++ less than a day
